src/scripts/helpers/formControllerHelpers.ts

```diff
@@ -0,0 +1,171 @@
+import type {
+  RenderData,
+  GetFormOptions,
+  PostFormOptions,
+  ValidationResult
+} from '#types/form-controller-types.js';
+
+const BAD_REQUEST = 400;
+const MINIMUM_FIELDS_LENGTH = 0;
+
+/**
+ * Generic function to handle GET requests for edit forms
+ * @param {Request} req - Express request object
+ * @param {Response} res - Express response object
+ * @param {NextFunction} next - Express next middleware function
+ * @param {GetFormOptions} options - Configuration options for the form handler
+ * @returns {Promise<void>}
+ */
+export async function handleGetEditForm(
+  req: Request,
+  res: Response,
+  next: NextFunction,
+  options: GetFormOptions
+): Promise<void> {
+  const caseReference = safeString(req.params.caseReference);
+  try {
+    const response = await apiService.getClientDetails(req.axiosMiddleware, caseReference);
+
+    let extractedData: Record<string, unknown> = {};
+    if (options.fieldConfigs !== undefined) {
+      // Use field configurations for data extraction
+      extractedData = extractCurrentFields(response.data, options.fieldConfigs);
+    } else if (options.dataExtractor !== undefined) {
+      // Use custom data extractor function
+      extractedData = options.dataExtractor(response.data);
+    }
+
+    const templateData = {
+      caseReference,
+      csrfToken: typeof req.csrfToken === 'function' ? req.csrfToken() : undefined,
+      ...extractedData
+    };
+    res.render(options.templatePath, templateData);
+  } catch (error) {
+    next(error);
+  }
+}
+
+/**
+ * Generic function to handle POST requests for edit forms
+ * @param {Request} req - Express request object
+ * @param {Response} res - Express response object
+ * @param {NextFunction} next - Express next middleware function
+ * @param {PostFormOptions} options - Configuration options for the form handler
+ * @returns {Promise<void>}
+ */
+export async function handlePostEditForm(
+  req: Request,
+  res: Response,
+  next: NextFunction,
+  options: PostFormOptions
+): Promise<void> {
+  const caseReference = safeString(req.params.caseReference);
+  const { templatePath, fields, apiUpdateData, useCustomValidation = false } = options;
+
+  let inputErrors: Record<string, string> = {};
+  let errorSummaryList: Array<{ text: string; href: string }> = [];
+  let formIsInvalid = false;
+
+  if (useCustomValidation) {
+    // Use express-validator validation
+    const validationErrors: Result<ValidationErrorData> = validationResult(req).formatWith(formatValidationError);
+
+    if (!validationErrors.isEmpty()) {
+      const resultingErrors = validationErrors.array().map((errorData: ValidationErrorData) => {
+        // Determine field name based on the validation error content
+        let fieldName = ''; // default to empty string
+
+        // Check for specific validation types
+        if (errorData.summaryMessage.toLowerCase().includes('postcode')) {
```

Comment:
Comment - not sure how to overcome this if/else statement but it does seem "brittle". The error summary messages may change and you'd have to remember to change here.  Having strings in `en.yml` might help this??

We could map it to an array?

> const keywordMap = [
>   { keyword: 'postcode', field: 'postcode' },
>   { keyword: 'update the client address', field: 'address' },
>   { keyword: 'phone', field: 'phoneNumber' }
> ];

Reply:
I have updated to get the fieldName from `express-validator` rather than parsing the error message.